### PR TITLE
Add a specialized FFT layer

### PR DESCRIPTION
**Pull request by benson31**
This PR adds support for a special flavor of FFT layer. Where a general FFT layer would take real- or complex-valued input and return complex-valued output (and handle complex-valued back-prop appropriately), the layer added here, termed `dft_abs_layer` for a balance between clarity and brevity, maintains the real-in-real-out paradigm by taking the entrywise absolute value of the output of the FFT. It handles backprop for this case as well.

This PR depends on LLNL/Elemental#106 as the Hydrogen support for complex numbers on GPUs was lacking. As it is, it's hacky but functional.

TODO:

- [x] Finish GPU backprop impl (waiting on a missing kernel in Hydrogen).
- [x] Test `dft_abs_layer<float, GPU>` (bamboo currently forces CPU-only).

### CODE REVIEW

**Comment by benson31** [comment]
This is ready to go once LLNL/Elemental#106 merges.

**Review by bvanessen — APPROVED** [review]
LGTM

File: src/utils/unit_test/fftw_test.cpp [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    {
+      T const* feat_map_mat = full_output.LockedBuffer()
+        + sample*full_output.LDim()
+        + feat_map*num_entries;
+      for (int ent = num_entries_r2c; ent < num_entries; ++ent)
+      {
+        auto const conj_ent = (ent == 0 ? 0 : num_entries-ent);
+        auto const val = feat_map_mat[ent];
+        auto const conj_val = feat_map_mat[conj_ent];
+        if (val != conj_val)
```
Comment (timmoon10) [inline src/utils/unit_test/fftw_test.cpp:87]: Shouldn't we be checking they are conjugates?

File: src/utils/unit_test/fftw_test.cpp [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        {
+          auto const conj_row = (row == 0 ? 0 : num_rows-row);
+          auto const conj_col = (col == 0 ? 0 : num_cols-col);
+          auto const idx = col + row*num_cols;
+          auto const conj_idx = conj_col + conj_row*num_cols;
+          auto const val = feat_map_mat[idx];
+          auto const conj_val = feat_map_mat[conj_idx];
+
+          if ((Approx(El::RealPart(val)) != El::RealPart(conj_val))
+              || false)//(Approx(El::ImagPart(val)) != -El::ImagPart(conj_val)))
```
Comment (timmoon10) [inline src/utils/unit_test/fftw_test.cpp:135]:
```suggestion
              || (Approx(El::ImagPart(val)) != -El::ImagPart(conj_val)))
```

File: src/utils/unit_test/fftw_test.cpp [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+  // Do some initializations
+  El::MakeUniform(mat, DataT(0.f), RealT(2.f));
+  El::Copy(mat, mat_orig);
+
+  // Compute the forward transformation
+  REQUIRE_NOTHROW(fftw.compute_forward(mat));
+
+  // Verify the results
+  //REQUIRE(assert_r2c_symmetry(mat, dims));
```
Comment (timmoon10) [inline src/utils/unit_test/fftw_test.cpp:376]:
```suggestion
  REQUIRE(assert_r2c_symmetry(mat, dims));
```

**Review by timmoon10 — APPROVED** [review]
Mostly LGTM. There seem to be a few problems in the FFTW test (probably fixed by copy-pasting the input data from the cuFFT test), but I don't see it as a huge priority.

File: src/utils/unit_test/fftw_test.cpp (hunk 1 above)
Comment (benson31) [reply src/utils/unit_test/fftw_test.cpp:87]: This was some debugging code that never got cleaned up. I'll do a quick tidy before merge.

File: bamboo/unit_tests/test_unit_layer_dft_abs.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    ])
+    return message
+
+# ==============================================
+# Setup PyTest
+# ==============================================
+
+# Create test functions that can interact with PyTest
+for test in tools.create_tests(setup_experiment, __file__):
+    globals()[test.__name__] = test
```
[the remark this answers is not in the document]
Comment (benson31) [reply bamboo/unit_tests/test_unit_layer_dft_abs.py:177]: It does.

File: src/utils/unit_test/fftw_test.cpp (hunk 1 above)
Comment (benson31) [reply src/utils/unit_test/fftw_test.cpp]: Done

**Pull request merged**